src/Grpc.Net.Client/Internal/GrpcCall.cs

```diff
@@ -202,163 +185,153 @@ public void EnsureNotDisposed()
             }
         }
 
-        public void EnsureHeadersValid()
-        {
-            if (_headerValidationError != null)
-            {
-                throw new InvalidOperationException(_headerValidationError);
-            }
-        }
-
-        public Exception CreateCanceledStatusException()
-        {
-            if (_headerValidationError != null)
-            {
-                return new InvalidOperationException(_headerValidationError);
-            }
-
-            var statusCode = DeadlineReached ? StatusCode.DeadlineExceeded : StatusCode.Cancelled;
-            return new RpcException(new Status(statusCode, string.Empty));
-        }
-
         /// <summary>
         /// Marks the response as finished, i.e. all response content has been read and trailers are available.
         /// Can be called by <see cref="GetResponseAsync"/> for unary and client streaming calls, or
         /// <see cref="HttpContentClientStreamReader{TRequest,TResponse}.MoveNextCore(CancellationToken)"/>
         /// for server streaming and duplex streaming calls.
         /// </summary>
-        public void FinishResponse()
+        public void FinishResponse(bool throwOnFail, Status? status = null)
         {
             ResponseFinished = true;
             Debug.Assert(HttpResponse != null);
 
             try
             {
-                // Get status from response before dispose
-                // This may throw an error if the grpc-status is missing or malformed
-                var status = GetStatusCore(HttpResponse);
-
-                if (status.StatusCode != StatusCode.OK)
+                if (status == null)
                 {
-                    Log.GrpcStatusError(Logger, status.StatusCode, status.Detail);
-                    throw new RpcException(status);
+                    if (!TryGetStatusCore(HttpResponse, out status))
+                    {
+                        status = new Status(StatusCode.Cancelled, "No grpc-status found on response.");
+                    }
                 }
             }
-            finally
+            catch (Exception ex)
             {
-                Log.FinishedCall(Logger);
+                status = new Status(StatusCode.Cancelled, ex.Message);
+            }
+
+            // Clean up call resources once this call is finished
+            // Call may not be explicitly disposed when used with unary methods
+            // e.g. var reply = await client.SayHelloAsync(new HelloRequest());
+            DisposeCore(status.Value);
```

Comment:
I have separated them so that there is a difference between "everything in this call is done, clean up all the resources" and "the user has explicitly called Dispose and now methods should throw `ObjectDisposedException`"